### ISSUE TITLE
Switch to sync model if queue is full

### ISSUE THREAD

**Issue by ilchanka** [airbrake/airbrake-ruby#188]
Is it possible to switch sync calls if I have my queue full? Currently if I will send 1000 errors and the queue is only 100, 900 will be ignored, really unexpected for me.

Looks like https://github.com/airbrake/airbrake-ruby/pull/47 this PR removed exactly what I I'm asking for)

Maybe it's better to have config option to switch between these 2 models(blocking or rejecting)? Looks like it's easy to add

**Comment by kyrylo** [comment]
It's easy to add but hard to maintain both :P 
Can you elaborate why this is unexpected?

We also allow configuring the queue with `queue_size`: https://github.com/airbrake/airbrake-ruby#queue_size
In my opinion it should be enough to cover your needs. Just set it to 1000 and you're good :)

**Comment by ilchanka** [comment]
Hi @kyrylo. It is unexpected because if I have 1001 errors raised I expect to see all of them in airbrake. Not 1000, if we would write some school project it is ok to lose some errors. But if we are working with enterprise and money, it is crucial to have exact numbers. 

For example today morning we had one of our external dependency services partially working. And in a few hours we got 33k errors. Having queue size at 1000 doesn't guarantee that all of them will be saved. Setting 10k can just blow memory usage in one time, ok, maybe not blow, but memory will definitely jump, so I'm aware using high size queue.

So for now we are not upgrading version of airbrake gem in some critical services only because we can lose some errors. Would be nice to have this option

**Comment by kyrylo** [comment]
>  if we would write some school project it is ok to lose some errors. But if we are working with enterprise and money, it is crucial to have exact numbers.

This is a universal argument that applies to many topics. It's always nice to have exact everything, but reality is there are always edge cases and limitations. We have to have trade-offs.

In theory, you can send 1M errors in a few hours from this notifier, but you will see that your project will be rate limited according to your [plan](https://airbrake.io/pricing). We also apply [some techniques to reduce error size](https://airbrake.io/docs/airbrake-faq/error-reduction-in-very-large-groups/) of errors with many occurrences. You already can lose occurrences even with your older version of airbrake-ruby.

I am not sure having a toggle for changing this behaviour is a good idea. This is already complex behaviour and having a toggle will only add up to the complexity. We should settle on a good default and in my opinion what we offer now is good enough (especially when you can configure queue size).

However, this decision wasn't imposed by rate limiting or the complexity. From technical implementation point of view, I agree with #47: [`send`](https://github.com/airbrake/airbrake-ruby/blob/2c1aced543181c565052e5e4c98c6f169f359861/lib/airbrake-ruby/async_sender.rb#L29) (what powers `Airbrake.notify`) should not block no matter what. It must be asynchronous as we advertise.

---

Since you mentioned #47, I'd like to CC @crunis who was the author of the change (in case he's interested in this discussion).

**Comment by crunis** [comment]
If there's a problem in our apps that is generating many errors there's little value logging them all, there's more value maximising our app's health by ensuring it doesn't get blocked by the error logging function that is sending usually the same repeated error.

If, in your case, every error has intrinsic value (maybe it holds some important information of the transaction, which doesn't look like a good idea to me), I would just stick to sending them synchronously. If you still want to benefit from asynchronous performance, add your own asynchronous logic in your app that falls back to synchronously behaviour in the fashion that best suits your purpose.

In brief: I think that is more annoyingly surprising to encounter that your app's performance is being heavily penalised because some asynchronous function all of a sudden is behaving synchronously than to miss some errors when you are receiving a ton of them.

**Comment by ilchanka** [comment]
> it doesn't get blocked by the error logging function that is sending usually the same repeated error.

That what I really was expecting when I first tried async version. And then was confused when I saw that the number of errors is smaller than I was sending in tests

> If, in your case, every error has intrinsic value (maybe it holds some important information of the transaction, which doesn't look like a good idea to me)

I'm more interested in number of requests.
Now we are using old version gem with sync model, and blocking is not a surprise, everything is blocked)) 

> If you still want to benefit from asynchronous performance, add your own asynchronous logic in your app that falls back to synchronously behaviour in the fashion that best suits your purpose.

I don't really like monkey patching. So there is another proposal. We can have config option to point to any sender adapter. So anyone can write own adapter. This will require a lot of work to implement, but will be more maintainable and flexible

**Comment by crunis** [comment]
Ok! Seems that we all agree that an async function shouldn't behave synchronously. Nice.

Regarding your proposal of a config option to point to any sender adapter, you already mention is probably a lot of work to implement and I would add that is also work for @kyrylo to maintain, I honestly don't think is worth the effort, if someone has very specific error handling needs, he should write his own error handling code with specific sync/async behaviour and simply use the airbrake gem to deliver the error.

If you want my particular opinion on your scenario, you say that you are "interested in the number of requests", if this is a common situation, to keep track of the exact number of requests that fail and do something with that number, I think you should find another approach with another tool, if this is just a sporadic need, remember that every time airbrake's asynchronous function drops an error [it writes it in the log](https://github.com/airbrake/airbrake-ruby/blob/master/lib/airbrake-ruby/async_sender.rb#L116-L121) , so the information is not really lost and you can always go to the logs and obtain it.

**Comment by kyrylo** [comment]
> @ilchanka 
> [...] And then was confused when I saw that the number of errors is smaller than I was sending in tests

You shouldn't be testing our library (we already have lots of tests 🍻), but I am not sure what kind of tests you are writing.

> We can have config option to point to any sender adapter. So anyone can write own adapter.

This sounds sensible, but that means giving a lot of freedom to users. Our current implementation is dead simple, and it's also easy to debug. In fact, I don't remember any recent issues with it. 

Your suggested approach opens a whole new wormhole: there are so many possible adapters, and there can be many bugs related (and unrelated) to Airbrake, many of which would need my assistance. If people configure an external adapter and say "My exception isn't reaching Airbrake, help me please", then I have no way to debug it anymore. 

Allowing external queue adapters is very powerful, I agree. The question is do we need that flexibility in this library? I am inclined to think that we should stay on the simpler side. Yes, we may lose occurrences, but as @crunis pointed out, they're logged. It's not perfect, but it's a good compromise, in my opinion.

---

> @crunis 

Thanks for chiming in!

---

@all 

I'm not saying a no to this feature, but as of now I see little interest in this feature and I see no enough evidence that this is truly needed.

**Comment by ilchanka** [comment]
> You shouldn't be testing our library (we already have lots of tests 🍻), but I am not sure what kind of tests you are writing.

I just made some manual/performance testing to ensure that everything works correctly

> but as @crunis pointed out, they're logged. It's not perfect, but it's a good compromise, in my opinion.

This is not really useful because I don't have notifications if something is really added there, and checking logs for something new every time is painful.

> I'm not saying a no to this feature, but as of now I see little interest in this feature and I see no enough evidence that this is truly needed.

I will write a wrapper to use in our services then. Thanks @kyrylo @crunis
